graylog2-server/src/main/java/org/graylog2/indexer/messages/SerializationMemoizingMessage.java

```diff
@@ -0,0 +1,181 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.indexer.messages;
+
+import com.codahale.metrics.Meter;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheStats;
+import jakarta.annotation.Nonnull;
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Message;
+import org.graylog2.plugin.utilities.ratelimitedlog.RateLimitedLogFactory;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+public class SerializationMemoizingMessage implements ImmutableMessage {
+    private static final Logger LOG = LoggerFactory.getLogger(SerializationMemoizingMessage.class);
+    private static final Logger RATE_LIMITED_LOG = RateLimitedLogFactory.createRateLimitedLog(
+            LOG, 1, Duration.ofSeconds(20));
+
+    private final Message delegate;
+
+    private final Cache<ObjectMapper, CacheEntry> serializedBytesCache;
+
+    private record CacheEntry(byte[] serializedBytes, Meter invalidTimeStampMeter) {}
+
+    public SerializationMemoizingMessage(Message delegate) {
+        this(delegate, LOG.isDebugEnabled());
+    }
+
+    public SerializationMemoizingMessage(Message delegate, boolean enableCacheStats) {
+        this.delegate = delegate;
+
+        final var cacheBuilder = CacheBuilder.newBuilder().weakKeys().softValues();
+        if (enableCacheStats) {
+            this.serializedBytesCache = cacheBuilder.recordStats().build();
+        } else {
+            this.serializedBytesCache = cacheBuilder.build();
+        }
+    }
+
+    @Override
+    public synchronized byte[] serialize(ObjectMapper objectMapper, @Nonnull Meter invalidTimestampMeter)
+            throws JsonProcessingException {
+
+        try {
+            final var cacheEntry = serializedBytesCache.get(objectMapper, () -> {
+                        final var tsMeter = new Meter();
+                        return new CacheEntry(delegate.serialize(objectMapper, tsMeter), tsMeter);
+                    }
+            );
+            invalidTimestampMeter.mark(cacheEntry.invalidTimeStampMeter().getCount());
+            if (RATE_LIMITED_LOG.isDebugEnabled()) {
+                if (cacheStats().evictionCount() > 0) {
+                    RATE_LIMITED_LOG.debug("The JVM cleared a cached serialized message because of memory pressure.");
+                }
+            }
```

Comment:
I think we can replace this with a `SoftReference<CacheEntry>` field that is either null (never been set, so we initialize it on first use) or non-null, in which case `.get()` is either null (evicted, log warning) or non-null (we have the cached value, just use that).

This approach would save quite a bit of memory over the Cache with a single entry, although the code is a bit uglier.
Of course we'd have not other stats in that implementation.

Reply:
If we don't go that route, we should see if we can minimize the cache size, i think it has a rather large initial capacity. And perhaps switch to Caffeine instead of Guava, but that's a minor comment.

Reply:
Thanks for the comment! I had what you suggested implemented in the first place, but with weak references instead of soft references, which was wrong (see https://github.com/Graylog2/graylog2-server/blob/9873989077b0060f3cdcdf97d82a000e25138dab/graylog2-server/src/main/java/org/graylog2/indexer/messages/SerializationMemoizingMessage.java). But then I thought I'd just use a guava cache so that it's a bit more maintainable for others. But there's really no need, and the initial cache size is a good hint!

I think I'll go back to what I had before and what you suggested. I'll probably add the object mapper to the memoized entry so that we don't return the wrong serialization if a different object mapper is used on consecutive calls. But that shouldn't be the norm and we don't need to cache per object mapper but replace our single memoized value in that case.

Reply:
the cache is gone